torch/csrc/jit/codegen/cuda/lower_utils.cpp

```diff
@@ -277,6 +337,195 @@ std::pair<kir::ForLoop*, int64_t> getAllocPoint(
+    }
+    replaced_expr_ = new_for_loop;
+  }
+
+  void visit(kir::IfThenElse* ite) final {
+    auto new_ite = ir_builder_.create<kir::IfThenElse>(ite->predicate());
+    auto replaced_then_body =
+        replace(ite->thenBody().exprs(), replacement_map_);
+    for (auto new_expr : replaced_then_body) {
+      new_ite->thenBody().push_back(new_expr);
+    }
+    if (ite->hasElse()) {
+      auto replaced_else_body =
+          replace(ite->elseBody().exprs(), replacement_map_);
+      for (auto new_expr : replaced_else_body) {
+        new_ite->elseBody().push_back(new_expr);
+      }
+    }
+    replaced_expr_ = new_ite;
+  }
+
+  void visit(kir::UnaryOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::UnaryOp>(
+          node->operation(),
+          node->out(),
+          replaced_inputs.value().at(node->in()));
+    }
+  }
+  void visit(kir::BinaryOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::BinaryOp>(
+          node->operation(),
+          node->out(),
+          replaced_inputs.value().at(node->lhs()),
+          replaced_inputs.value().at(node->rhs()));
+    }
+  }
+
+  void visit(kir::TernaryOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::TernaryOp>(
+          node->operation(),
+          node->out(),
+          replaced_inputs.value().at(node->in1()),
+          replaced_inputs.value().at(node->in2()),
+          replaced_inputs.value().at(node->in3()));
+    }
+  }
+
+  void visit(kir::ReductionOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::ReductionOp>(
+          node->operation(),
+          node->init(),
+          node->out(),
+          replaced_inputs.value().at(node->in()));
+    }
+  }
+
+  void visit(kir::BroadcastOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::BroadcastOp>(
+          node->out(), replaced_inputs.value().at(node->in()));
+    }
+  }
+
+  void visit(kir::WelfordOp* node) final {
+    auto replaced_inputs = getMaybeInputReplacementMap(node);
+    if (replaced_inputs.has_value()) {
+      replaced_expr_ = ir_builder_.create<kir::WelfordOp>(
+          node->outAvg(),
+          node->outVar(),
+          node->outN(),
+          node->initAvg(),
+          node->initVar(),
+          node->initN(),
+          replaced_inputs.value().at(node->inAvg()),
+          replaced_inputs.value().at(node->inVar()),
+          replaced_inputs.value().at(node->inN()));
+    }
+  }
```

Comment:
Sharing the same expression with multiple scopes doesn't seem really right to me. That means that `kir::Expr::scope_` doesn't reliably identify which scope the expression appears. It would not be a problem if traversed from scopes to expressions, but then we should get rid of the `scope_` field of `Expr` as it's broken.

Disallowing mutation entirely seems too restrictive to me. It would mean that just replacing one KIR node would need to copy the whole nodes, which seems rather inefficient.

Reply:
We should file an issue and address this outside this PR.

Reply:
Issue to track and address in a follow up: 
[https://github.com/csarofeen/pytorch/issues/1063](https://github.com/csarofeen/pytorch/issues/1063)